src/Coordination/KeeperDispatcher.cpp

```diff
@@ -610,6 +769,78 @@ void KeeperDispatcher::updateConfigurationThread()
     }
 }
 
+void KeeperDispatcher::finalizeRequests(const KeeperStorage::RequestsForSessions & requests_for_sessions)
+{
+    std::unordered_map<int64_t, size_t> counts_for_session;
+
+    for (const auto & request_for_session : requests_for_sessions)
+    {
+        ++counts_for_session[request_for_session.session_id];
+    }
+
+    std::lock_guard lock{unprocessed_request_mutex};
+    for (const auto [session_id, count] : counts_for_session)
+    {
+        auto unprocessed_requests_it = unprocessed_requests_for_session.find(session_id);
+        if (unprocessed_requests_it == unprocessed_requests_for_session.end())
+            continue;
+
+        auto & unprocessed_requests = unprocessed_requests_it->second;
+        unprocessed_requests.unprocessed_num -= count;
+
+        if (unprocessed_requests.unprocessed_num == 0)
+        {
+            if (!unprocessed_requests.request_queue.empty())
```

Comment:
Why nested if instead of `unprocessed_requests.unprocessed_num == 0 && !unprocessed_requests.request_queue.empty()`?

Reply:
because of the else below